src/sync/mpsc/mod.rs

```diff
@@ -0,0 +1,693 @@
+                // the configured buffer size
+                let park_self = match self.inner.buffer {
+                    Some(buffer) => state.num_messages > buffer,
+                    None => false,
+                };
+
+                // Only unpark the receive half if transitioning from 0 -> 1.
+                let unpark_recv = state.num_messages == 1;
+
+                return Some((park_self, unpark_recv));
+            }
+
+            curr = actual;
+        }
+    }
+
+    fn park(&mut self, can_park: bool) {
+        // TODO: clean up internal state if the task::park will fail
+
+        let task = if can_park {
+            Some(task::park())
+        } else {
+            None
+        };
+
+        self.maybe_parked = true;
+        *self.sender_task.lock().unwrap() = task;
+
+        // Send handle over queue
+        let t = self.sender_task.clone();
+        self.inner.parked_queue.push(t);
+    }
+}
+
+impl<T> Sink for Sender<T> {
+    type SinkItem = T;
+    type SinkError = SendError<T>;
+
+    fn start_send(&mut self, msg: T) -> StartSend<T, SendError<T>> {
+        self.start_send2(msg)
+    }
+
+    fn poll_complete(&mut self) -> Poll<(), SendError<T>> {
+        Ok(Async::Ready(()))
+    }
+}
+
+impl<T> Clone for Sender<T> {
+    fn clone(&self) -> Sender<T> {
+        self.try_clone().expect("failed to clone sender; maximum number of senders already reached")
+    }
+}
+
+impl<T> Drop for Sender<T> {
+    fn drop(&mut self) {
+        // Ordering between variables don't matter here
+        let prev = self.inner.num_senders.fetch_sub(1, Ordering::Relaxed);
+
+        if prev == 1 {
+            let _ = self.do_send(None, false);
+        }
+    }
+}
+
+/*
+ *
+ * ===== impl Receiver =====
+ *
+ */
+
+impl<T> Receiver<T> {
+    /// Closes the receiving half
+    ///
+    /// This prevents any further messages from being sent on the channel while
+    /// still enabling the receiver to drain messages that are buffered.
+    pub fn close(&mut self) {
+        // A relaxed memory ordering is acceptable given that toggling the
+        // flag is an isolated operation. If no further functions are
+        // called on `Receiver` then the outcome of this function doesn't
+        // really matter. If `poll` is called after this, then the the same
+        // cell will be operated on again with stronger ordering.
+        let mut curr = self.inner.state.load(Ordering::Relaxed);
+
+        loop {
+            let mut state = decode_state(curr);
+
+            if !state.is_open {
+                return;
+            }
+
+            state.is_open = false;
+
+            let next = encode_state(&state);
+            let actual = self.inner.state.compare_and_swap(curr, next, Ordering::Relaxed);
+
+            if actual == curr {
+                return;
+            }
+
+            curr = actual;
+        }
+    }
+
+    fn next_message(&self) -> Async<Option<T>> {
```

Comment:
For this function to be safe (the `message_queue.pop` below) could this be tagged `&mut self`?